tutorials/dataframe/df103_NanoAODHiggsAnalysis.C

```diff
@@ -0,0 +1,517 @@
+   auto df_el_track = df_el_sip3d.Filter("All(Electron_sip3d_el<4) && All(abs(Electron_dxy)<0.5) && "
+                                         "All(abs(Electron_dz)<1.0)",
+                                         "Electron track close to primary vertex with small uncertainty");
+   auto df_mu_ip3d = df_el_track.Define("Muon_ip3d_mu", "sqrt(Muon_dxy*Muon_dxy + Muon_dz*Muon_dz)");
+   auto df_mu_sip3d = df_mu_ip3d.Define("Muon_sip3d_mu", "Muon_ip3d_mu/sqrt(Muon_dxyErr*Muon_dxyErr + "
+                                                         "Muon_dzErr*Muon_dzErr)");
+   auto df_mu_track = df_mu_sip3d.Filter("All(Muon_sip3d_mu<4) && All(abs(Muon_dxy)<0.5) && "
+                                         "All(abs(Muon_dz)<1.0)",
+                                         "Muon track close to primary vertex with small uncertainty");
+   auto df_2p2n = df_mu_track.Filter("Sum(Electron_charge)==0 && Sum(Muon_charge)==0",
+                                     "Two opposite charged electron and muon pairs");
+   return df_2p2n;
+}
+
+// Reconstruct two Z candidates from four leptons of the same kind
+RVec<RVec<size_t>> reco_zz_to_4l(rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass, rvec_i charge)
+{
+   RVec<RVec<size_t>> idx(2);
+
+   // Find first lepton pair with invariant mass closest to Z mass
+   const auto z_mass = 91.2;
+   auto idx_cmb = Combinations(pt, 2);
+   auto best_mass = -1;
+   size_t best_i1 = 0;
+   size_t best_i2 = 0;
+   for (size_t i = 0; i < idx_cmb[0].size(); i++) {
+      const auto i1 = idx_cmb[0][i];
+      const auto i2 = idx_cmb[1][i];
+      if (charge[i1] != charge[i2]) {
+         TLorentzVector p1, p2;
+         p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+         p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+         const auto this_mass = (p1 + p2).M();
+         if (std::abs(z_mass - this_mass) < std::abs(z_mass - best_mass)) {
+            best_mass = this_mass;
+            best_i1 = i1;
+            best_i2 = i2;
+         }
+      }
+   }
+   idx[0].push_back(best_i1);
+   idx[0].push_back(best_i2);
+
+   // Reconstruct second Z from remaining lepton pair
+   for (size_t i = 0; i < 4; i++) {
+      if (i != best_i1 && i != best_i2) {
+         idx[1].push_back(i);
+      }
+   }
+
+   // Return indices of the pairs building two Z bosons
+   return idx;
+}
+
+// Compute Z masses from four leptons of the same kind and sort ascending in distance to Z mass
+RVec<float> compute_z_masses_4l(const RVec<RVec<size_t>> &idx, rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass)
+{
+   RVec<float> z_masses(2);
+   for (size_t i = 0; i < 2; i++) {
+      TLorentzVector p1, p2;
+      const auto i1 = idx[i][0];
+      const auto i2 = idx[i][1];
+      p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+      p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+      z_masses[i] = (p1 + p2).M();
+   }
+   const auto z_mass = 91.2;
+   if (std::abs(z_masses[0] - z_mass) < std::abs(z_masses[1] - z_mass)) {
```

Comment:
Perhaps we need a professor, but maybe is it safe to assume that the on shell is always much more massive than the offshell?